shell/platform/linux/fl_renderer_wayland.cc

```diff
@@ -0,0 +1,322 @@
+// (which blocks for a round trip to the Wayland compositor).
+// Subsequent calls return immediately.
+static void lazy_initialize_wayland_globals() {
+  if (registry_global) {
+    return;
+  }
+
+  GdkWaylandDisplay* gdk_display =
+      GDK_WAYLAND_DISPLAY(gdk_display_get_default());
+  g_return_if_fail(gdk_display);
+
+  wl_display* display = gdk_wayland_display_get_wl_display(gdk_display);
+  registry_global = wl_display_get_registry(display);
+  wl_registry_add_listener(registry_global, &registry_listener, nullptr);
+  wl_display_roundtrip(display);
+}
+
+G_DEFINE_TYPE(FlRendererWayland, fl_renderer_wayland, fl_renderer_get_type())
+
+// Returns the GTK widget at the top of the view's widget hierarchy.
+static GtkWidget* fl_renderer_wayland_get_toplevel_widget(
+    FlRendererWayland* self) {
+  FlRenderer* renderer = FL_RENDERER(self);
+  return gtk_widget_get_toplevel(GTK_WIDGET(fl_renderer_get_view(renderer)));
+}
+
+// Implements GObject::dispose.
+static void fl_renderer_wayland_dispose(GObject* object) {
+  FlRendererWayland* self = FL_RENDERER_WAYLAND(object);
+
+  GtkWidget* toplevel = fl_renderer_wayland_get_toplevel_widget(self);
+  g_signal_handlers_disconnect_by_func(
+      G_OBJECT(toplevel), (gpointer)fl_renderer_wayland_on_window_map, self);
+  g_signal_handlers_disconnect_by_func(
+      G_OBJECT(toplevel), (gpointer)fl_renderer_wayland_on_window_unmap, self);
+
+  g_clear_pointer(&self->subsurface.subsurface, wl_subsurface_destroy);
+  g_clear_pointer(&self->subsurface.egl_window, wl_egl_window_destroy);
+  g_clear_pointer(&self->subsurface.surface, wl_surface_destroy);
+
+  g_clear_pointer(&self->resource.egl_window, wl_egl_window_destroy);
+  g_clear_pointer(&self->resource.surface, wl_surface_destroy);
+
+  G_OBJECT_CLASS(fl_renderer_wayland_parent_class)->dispose(object);
+}
+
+// Implements FlRenderer::create_display.
+static EGLDisplay fl_renderer_wayland_create_display(FlRenderer* /*renderer*/) {
+  GdkWaylandDisplay* gdk_display =
+      GDK_WAYLAND_DISPLAY(gdk_display_get_default());
+  g_return_val_if_fail(gdk_display, nullptr);
+  return eglGetDisplay(gdk_wayland_display_get_wl_display(gdk_display));
+}
+
+static void fl_renderer_wayland_on_window_map(GtkWidget* widget,
+                                              FlRendererWayland* self) {
+  if (self->subsurface.subsurface) {
+    g_error("fl_renderer_wayland_on_window_map: already has a subsurface");
+    return;
+  }
+
+  GdkWaylandDisplay* gdk_display =
+      GDK_WAYLAND_DISPLAY(gdk_display_get_default());
+
+  wl_compositor* compositor =
+      gdk_wayland_display_get_wl_compositor(gdk_display);
+
+  lazy_initialize_wayland_globals();
+  if (!subcompositor_global) {
+    g_error(
+        "fl_renderer_wayland_on_window_map: could not bind to "
+        "wl_subcompositor");
+    return;
+  }
+
+  GdkWaylandWindow* window = GDK_WAYLAND_WINDOW(
+      gtk_widget_get_window(fl_renderer_wayland_get_toplevel_widget(self)));
```

Comment:
The `fl_renderer_wayland_get_toplevel_widget(self)` call is redundant right? Isn't this just `widget`?

Reply:
No. The view widget is not the toplevel widget, and the Wayland renderer *does* need to connect to the toplevel's map/unmap signals in order to properly handle the window being hidden and shown.

Reply:
I tested this locally to confirm, and `widget` is the same as `fl_renderer_wayland_get_toplevel_widget(self)` (because that's the object you subscribed to the map signal on).

I've got it working without the need to store view here - https://github.com/wmww/flutter_engine/pull/1. I think with those changes this branch should be good to land.

Reply:
Oh, right, you're talking about the widget argument to the callback. I guess that would be the same. Sorry, I misunderstood but that makes sense.